Scan unchecked resource-types

EstebanFS opened the pull request:
## What does this PR accomplish?
**Bug Fix** | Feature | Documentation

closes #6521

## Changes proposed by this PR:
This covers the edge case where a put-only resource's parent-type would not be check because we were only looping over everything except put-only resources

## Notes to reviewer:
If you need extra help reproducing the issue, you can check Alex's comment https://github.com/concourse/concourse/issues/6521#issuecomment-813596246

## Release Note
* Fixed an edge case where a put-only resource's parent-type would not be checked

## Contributor Checklist
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [ ] ~Updated [Documentation]~
- [x] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).


cc @taylorsilva

EstebanFS committed:
Scan unchecked resource-types

This covers the edge case where a put-only resource parent-type would not be check, because we were only looping over everything except put-only resources

Signed-off-by: Taylor Silva <tsilva@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw commented on atc/lidar/scanner.go:
```diff
@@ -46,6 +46,17 @@ func (s *scanner) Run(ctx context.Context) error {
 	resourceTypesChecked := &sync.Map{}
 
+	s.scanResources(spanCtx, resources, resourceTypes, waitGroup, resourceTypesChecked)
+	waitGroup.Wait()
```
Is there a reason we want the two sets of "scans" to run sequentially? Or can we get rid of this `waitGroup.Wait()`?

taylorsilva replied:
This gave me a different idea. Got rid of the `sync.Map` and put the `sync.WaitGroup` inside the `scan*` funcs.

aoldershaw replied:
That's kind of what I had in mind as well, but wasn't totally convinced it would work - e.g. consider the following:

```yaml
resource_types:
- name: grandparent
  type: registry-image
- name: parent
  type: grandparent

resources:
- name: child
  type: parent
```

If we don't perform the recursive `s.check` for parent types, it's possible that we create a check for `parent` before its parent type `grandparent` - might that be an issue? (genuinely not sure, but it feels odd)

taylorsilva replied:
Didn't even think about that use-case because I don't _think_ someone would use anything else besides `registry-image` as the `type` for `resource_types`. I think this would cause some unexpected behaviour depending on the order of the `resourceTypes` returned by the db. If `grandparent` appears before `parent` then everything is fine. If not, then for **one check interval**, an older version of `grandparent` would be used to fetch `parent`, instead of the most recently available version of `grandparent`. This would immediately be resolved on the next check for these resource-types though. Maybe this is fine? We _eventually_ get you to the expected result? wdyt? I'm a bit on the fence about this.

aoldershaw replied:
I guess what's not clear to me is what `s.check` actually does. It doesn't actually run the check - it just creates a build in the DB (if it needs to) for the build tracker to pick up in order to actually run the check.

So, I *think* it shouldn't matter - I believe an older version of `grandparent` will always be used to fetch `parent` for a single check interval, even if we create the build for `grandparent` first

taylorsilva replied:
Good point. I haven't tested this myself but I'm not even sure if there's a max-in-flight for all checks. I remember a PR about smoothing out the number of checks getting run; I forget where that is.
I'm fine with this PR as-is then. Up to you if you think we should still nudge this back to the recursive `s.check` or keep this simpler one.

aoldershaw replied:
There's some rate limiting for running checks, but that's done within the check step itself (the check delegate, to be more precise):
https://github.com/concourse/concourse/blob/429eead503de082bd1eac1211c7afd0ed20418dc/atc/engine/check_delegate.go#L81

...however, that doesn't seem to rate limit resource types, just resources 🤔

In any event, I think the simplified code is nicer, I'm happy with it 👍

aoldershaw commented on atc/lidar/scanner.go:
```diff
@@ -58,13 +69,30 @@ func (s *scanner) Run(ctx context.Context) error {
-	return nil
+		go func(resourceType db.ResourceType, resourceTypes db.ResourceTypes) {
+			defer func() {
+				err := util.DumpPanic(recover(), "scanning resource type %d", resourceType.ID())
```
A bit confused about when this could panic. I guess it's just being defensive so as not to take down the ATC due to a programming mistake - but feels like that should get caught in tests, as it's just relatively simple control flow to create a build in the DB.

p.s. I recognize this panic recovery was present in the scanner beforehand - happy to keep it, just feels a bit odd to me

taylorsilva replied:
I say we keep it. Came from this PR: https://github.com/concourse/concourse/pull/5842

aoldershaw commented on atc/lidar/scanner.go:
```diff
@@ -58,13 +69,30 @@ func (s *scanner) Run(ctx context.Context) error {
+		waitGroup.Add(1)
 
-	return nil
+		go func(resourceType db.ResourceType, resourceTypes db.ResourceTypes) {
```
Minor thing, but we don't need to create a goroutine when the resource type is already checked. e.g. this could be:

```go
for _, resourceType := range resourceTypes {
	if _, exists := resourceTypesChecked.LoadOrStore(resourceType.ID(), true); exists {
		// check has already been created for resource type
		continue
	}
	waitGroup.Add(1)

	go func(resourceType db.ResourceType, resourceTypes db.ResourceTypes) {
		defer func() {
			err := util.DumpPanic(recover(), "scanning resource type %d", resourceType.ID())
			if err != nil {
				logger.Error("panic-in-scanner-run", err)
			}
		}()
		defer waitGroup.Done()

		s.check(ctx, resource, resourceTypes, resourceTypesChecked)
	}(resourceType, resourceTypes)
}
```

aoldershaw commented on atc/lidar/scanner.go:
```diff
@@ -46,6 +46,17 @@ func (s *scanner) Run(ctx context.Context) error {
 
+	s.scanResources(spanCtx, resources, resourceTypes, waitGroup, resourceTypesChecked)
+	waitGroup.Wait()
+	s.scanUncheckedResourceTypes(spanCtx, resourceTypes, waitGroup, resourceTypesChecked)
```
Now that we check all resource types, does `s.check` still need to check the parent resource type? Or perhaps the order the checks are created in matters

taylorsilva replied:
I think it still makes sense for `s.check` to check the parent resource-type first before running the resource's check. It feels out of order to run a check for a resource and then check if it's base resource-type has updated. I'm going to leave this as-is.

taylorsilva committed:
move waitGroup inside scan funcs

checks for all resource-types are performed first. Then checks for the
resources are performed. This PR simplifies the logic, removing the need
for a sync.Map. `s.check()` no longer needs to call itself to create a
check for the parent resource-type as well.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw commented on atc/lidar/scanner_test.go:
```diff
@@ -168,6 +168,23 @@ var _ = Describe("Scanner", func() {
+						Expect(fakeCheckFactory.TryCreateCheckCallCount()).To(Equal(2))
+
+						By("first check is for the default resource")
+						_, checkable, _, _, manuallyTriggered := fakeCheckFactory.TryCreateCheckArgsForCall(1)
```
This test is failing https://ci.concourse-ci.org/builds/142874260#L6087a117:1177, should this be:
```suggestion
						_, checkable, _, _, manuallyTriggered := fakeCheckFactory.TryCreateCheckArgsForCall(0)
```

taylorsilva committed:
Fix test to match new scanner behaviour

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw reviewed (approved):
LGTM

clarafu commented:
@taylorsilva Hey I noticed something small that we might want to address! Previously, we only creates checks for resource types when it is used by a resource and we only checked for resources in unpaused pipelines.

https://github.com/concourse/concourse/blob/08612f6c1a099176cb13536584ea074fd36057e5/atc/db/check_factory.go#L142-L144

But now that we create a check for all resource types, I think we probably want to continue to not create a check for resource types in `paused` pipelines. I think we just need to add a small`where` clause in this `checkFactory.ResourceTypes()` query

https://github.com/concourse/concourse/blob/08612f6c1a099176cb13536584ea074fd36057e5/atc/db/check_factory.go#L181-L183

taylorsilva commented:
Ah great catch! I'll PR and test it 👍 Thanks for call out!
